File: 11-container-with-most-water/container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int start = 0, end = height.length - 1;
        int maxWater = -1;
        while(start < end) {
            int minHeight = Math.min(height[start], height[end]);
            int distance = end - start;
            maxWater = Math.max(maxWater, distance * minHeight);
            while(start<end && height[start]<=minHeight) start++;
            while(start<end && height[end]<=minHeight) end--;
        }

        return maxWater;
        
    }
}

File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    # Two-pointer approach: start from both ends and calculate area using the shorter height; move the pointer at the shorter line inward to explore potentially larger areas, skipping heights ≤ current min since they can't improve the result.
    def maxArea(self, height: List[int]) -> int:
        start, end = 0, len(height) - 1
        maxLeft, maxRight = height[start], height[end]
        result = 0
        while(start < end):
            distance = end - start
            minHeight = min(height[start], height[end])
            result = max(result, distance * minHeight)
            while(start < end and height[start]<=minHeight): start+=1            
            while(start < end and height[end]<= minHeight): end-=1

        return result